modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBString.java

```diff
@@ -17,32 +17,38 @@
 
 package org.geotools.mbstyle.expression;
 
-import org.geotools.factory.CommonFactoryFinder;
 import org.geotools.mbstyle.parse.MBFormatException;
 import org.geotools.mbstyle.parse.MBObjectParser;
 import org.json.simple.JSONArray;
-import org.opengis.filter.FilterFactory2;
 import org.opengis.filter.expression.Expression;
+import org.opengis.filter.expression.Function;
 
-public class MBString extends MBExpression{
+import java.util.ArrayList;
+import java.util.List;
 
+public class MBString extends MBExpression{
+    MBObjectParser parse;
     public MBString(JSONArray json){
         super(json);
+        parse = new MBObjectParser(MBString.class);
     }
 
-
     /**
      * Returns a string consisting of the concatenation of the inputs.
      * Example: ["concat", string, string, ...]: string
      * @return
      */
     public Expression stringConcat(){
-        String concatString = "";
+        List<Expression> list = new ArrayList<>();
         for(int i =1; i < json.size(); i++) {
-            String s = (String) json.get(i);
-            concatString += s;
+            Object o = parse.get(json, i);
+            Expression e = ff.literal(o.toString());
```

Comment:
If `o` is an Expression (e.g. another function), I think it needs to remain as one, and should only be evaluated at render time, so I think the `toString()` call is an error here. It will work fine if you are just concating / upcasing character values, but as soon as you start dealing with properties coming from the underlying data (e.g. place names) it won't work.